src/Symfony/Component/Messenger/Transport/Dsn.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport;
+
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class Dsn
+{
+    private $scheme;
+    private $host;
+    private $user;
+    private $password;
+    private $port;
+    private $path;
+    private $options;
+
+    public function __construct(string $scheme, string $host, ?string $user = null, ?string $password = null, ?int $port = null, ?string $path = null, array $options = [])
+    {
+        $this->scheme = $scheme;
+        $this->host = $host;
+        $this->user = $user;
+        $this->password = $password;
+        $this->port = $port;
+        $this->path = $path;
+        $this->options = $options;
+    }
+
+    public static function fromString(string $dsn, array $options): self
+    {
+        if (false === $parsedDsn = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The "%s" messenger DSN is invalid.', $dsn));
+        }
+
+        $user = isset($parsedDsn['user']) ? urldecode($parsedDsn['user']) : null;
+        $password = isset($parsedDsn['pass']) ? urldecode($parsedDsn['pass']) : null;
+        $port = $parsedDsn['port'] ?? null;
+        $path = $parsedDsn['path'] ?? null;
+        parse_str($parsedDsn['query'] ?? '', $options);
+
+        return new self($parsedDsn['scheme'], $parsedDsn['host'], $user, $password, $port, $path, $options);
+    }
+
+    public function getScheme(): string
+    {
+        return $this->scheme;
+    }
+
+    public function getHost(string $default = null): ?string
```

Comment:
given return type may be null, wouldn't it be simpler to remove the default and let the caller performing the ternary? `$dsn->getHost() ?? 'default'`
just asking,

Reply:
IMHO `$dsn->getHost('default')` looks a little bit nicer than `$dsn->getHost() ?? 'default'`

Reply:
from an analysis pov `$dsn->getHost('default')` is still nullable :) so you still have to account for e.g. `$dsn->getHost('default') ?? 'absolute default?'`

 as such i agree with @jderusse `$dsn->getHost() ?? 'default'` is simpler

Reply:
SCA sounds reasonable, will do change it